README.md
## Yandex Disk gem
[![Code Climate](https://codeclimate.com/github/denwwer/yandex_disk.png?t=1)](https://codeclimate.com/github/denwwer/yandex_disk)
[![Code Coverage](https://codeclimate.com/github/denwwer/yandex_disk/coverage.png)](https://codeclimate.com/github/denwwer/yandex_disk)
[![security](https://hakiri.io/github/denwwer/yandex_disk/master.svg)](https://hakiri.io/github/denwwer/yandex_disk/master)

An easy-to-use pure Ruby client library for Yandex Disk API based on Net::HTTP. No any runtime dependencies required.

Has any issue or suggestion please write about it [here](https://github.com/denwwer/yandex_disk/issues).
## Usage
Include to project and login

    # require directly
    require 'yandex_disk'
    
    # or in Gemfile
    gem 'yandex_disk'
    
    # login
    yd = YandexDisk.login(login, pwd)
#### Upload file

    yd.upload(file, path, options)
return true if success else raise RequestError

**options:**

* `:chunk_size` file chunk size, default is 1024

* `:force` create path if not exist, default raise error


**example:**

    # create "my/work" directory and upload file to "/my/work/graph.pdf" using chunk size 500 per request
    yd.upload('/home/graph.pdf', 'my/work', { :force => true, :chunk_size => 500 })

#### Download file

    yd.download(file, save_path)

if `save_path` not present - return file content else return true when save file

**example:**
  
    # download file to "/home/graph.pdf"
    yd.download('/my/work/graph.pdf', '/home') #=> true
    
    # keep file content in memory
    yd.download('/my/work/graph.pdf') #=> String
#### Create path

    yd.create_path(path)

return true if success else raise RequestError

  **alias:** `mkdir`

  **example:**

    # create "/my/work" path
    yd.mkdir('/my/work')
#### Get available and used space

    yd.size(options)

  **options:**

  * `:h_size` return size in human readable format (for default return in bytes) 15 Byte, 100 KB, etc

  **example:**

    yd.size # => { :available => 213133, :used => 321 }
    yd.size(:h_size => true) #=> { :available =>  '7.8 GB', :used => '200 MB' }
#### File exist?

    yd.exist?(file) 

return true if file exist else false

  **example:**

    yd.exist?('/home/graph.pdf') #=> true
#### Get file properties

    yd.properties(file, options)

return Hash `{ :name, :created, :updated, :type, :size, :is_file, :public_url }` if success else raise RequestError

 **options:**

   * `:h_size` return size in human readable format (for default return in bytes) 15 Byte, 100 KB, etc

  **example:**

    prop = yd.properties('/home/graph.pdf', { :h_size => true })
    prop[:is_file] #=> true
    prop[:size] #=> 25 MB
#### Return list of files properties in directory

    yd.files(path, options)

  return Array with Hash `[{ :name, :created, :updated, :type, :size, :is_file }]` if success else raise RequestError

 **options:**

   * `:h_size` return size in human readable format (for default return in bytes) 15 Byte, 100 KB, etc

   * `:root` include properties for root directory ("home" in example)

  **alias:** `ls`

  **example:**

    files = yd.files('/home')
    files[0][:name] #=> graph.pdf
    
    # with root
    files = yd.properties('/home', { :root => true })
    files[0][:name] #=> 'home'
#### Copy file or directory

    yd.copy(from, to)

return true if successful else raise RequestError

  **alias:** `cp`

  **example:**

    yd.copy('/home/graph.pdf', '/backup') # copy "graph.pdf" to "backup" directory
#### Move file or directory

    yd.move(from, to)

return true if successful else raise RequestError

  **alias:** `mv`

  **example:**

    yd.move('/home/graph.pdf', '/pdfs') # move "graph.pdf" to "pdfs" directory
#### Delete file or directory

    yd.delete(path)

return true if successful else raise RequestError

  **alias:** `del`

  **example:**

    yd.delete('/home/graph.pdf') # delete "graph.pdf"
#### Set public access for file or directory

    yd.set_public(path)

return public url if successful else raise RequestError

  **example:**

    yd.set_public('/home/graph.pdf') #=> http://yadi.sk/d/#############
#### Set private file or directory

    yd.set_private(path)

return true if successful else false

  **example:**

    yd.set_private('/home/graph.pdf') #=> true
#### Get image preview

    preview(path, size, save_to)

save image if successful else raise RequestError
 size supported value:

    T-shirt size (like in Yandex.Fotki), such as size=M. Yandex.Disk returns a preview in the size you selected:
    XXXS — 50 pixels on each side (square).
    XXS  — 75 pixels on each side (square).
    XS   — 100 pixels on each side (square).
    S    — 150 pixels wide, preserves original aspect ratio.
    M    — 300 pixels wide, preserves original aspect ratio.
    L    — 500 pixels wide, preserves original aspect ratio.
    XL   — 800 pixels wide, preserves original aspect ratio.
    XXL  — 1024 pixels wide, preserves original aspect ratio.
    XXXL — 1280 pixels wide, preserves original aspect ratio.

    A number, such as size=128.
    Yandex.Disk returns a preview with this width. If the specified width is more than 100 pixels, the preview preserves the aspect ratio of the original image.
    Otherwise, the preview is additionally modified: the largest possible square section is taken from the center of the image to scale to the specified width.

    Exact dimensions, such as size=128x256.
    Yandex.Disk returns a preview with the specified dimensions. The largest possible section with the specified width/height ratio is taken from the center of the original image (in the example, the ratio is 128/256 or 1/2). Then this section of the image is scaled to the specified dimensions. See the example with exact dimensions below.

    Exact width or height, such as size=128x or size=x256.
    Yandex.Disk returns a preview with the specified width or height that preserves the aspect ratio of the original image.
*View the original [api page](http://api.yandex.com/disk/doc/dg/reference/preview.xml) for details.*

  **example:**
    
    yd.preview('/photo/car.jpg', 'm', '/home') # save "car.jpg" with 300 pixels wide to home directory
    yd.preview('/photo/car.jpg', 128, '/home') # save preview 128 pixels wide to home directory


ChangeLog.md
## ChangeLog

### v1.1.3
* fixed size for `files` method
* `download` method return `String` file raw data, if save path don't present

        yd.download('/my/work/graph.pdf') #=> String

* minor code fixing

### v1.1.2
* fixed Ruby 2.0 file upload error
* fixed Ruby 2.0 openssl error

### v1.1.1
* gem size decrease

### v1.1.0 (yanked)

* `size`, `properties`, `files` methods now has option `:h_size => true` (humanize size) to return readable format: 15 Byte, 100 KB, 128 MB, 1 GB.

        s = yd.size(:h_size => true)
        p s[:available], s[:used] => 7.9 GB, 100 MB
        ####
        prop = yd.properties('/home/graph.pdf', {:h_size => true})
        p prop[:size] # => 50 MB
        ###
        files = yd.files('/home', {:h_size => true})
        p files[0][:size] => 50 MB

* Method `files` changed second argument, now is `Hash`, and dont include root directory for default:

         # Old
         # dont include root directory
         yd.files('/home', false)
         # include root directory
         yd.files('/home')
         
         # New
         # dont include root directory
         yd.files('/home')
         # include root directory
         yd.files('/home', {:root => true})

* added alias method `ls` to `files`:

         yd.ls('/home')


spec/spec_helper.rb
# encoding: UTF-8
require 'codeclimate-test-reporter'
require 'simplecov'

SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
    SimpleCov::Formatter::HTMLFormatter,
    CodeClimate::TestReporter::Formatter
]
SimpleCov.start
require 'rubygems'
require 'yandex_disk'
require 'fileutils'

RSpec.configure do |config|
  config.fail_fast = true

  config.before(:each) do
    init_test_files
  end

  config.after(:each) do
    clear!
  end
end

# login and password for yandex disk
LOGIN = ''
PWD   = ''

# test settings
FILES_PATH    = File.dirname(__FILE__) + '/files'
DOWNLOAD_PATH = FILES_PATH + '/downloads'
FILE_TEXT     = 'Hi developer.'

def init_test_files
  FileUtils.mkdir_p(DOWNLOAD_PATH) if !Dir.exist?(DOWNLOAD_PATH)
  FileUtils.mkdir_p(FILES_PATH) if !Dir.exist?(FILES_PATH)

  @text_file = FILES_PATH + '/sample.txt'
  File.open(@text_file, 'w+'){|f| f.write(FILE_TEXT)} unless File.file? @text_file
  @text_file.freeze

  @image_file = FILES_PATH + '/sample.jpg'
  @image_file.freeze

  raise Exception, "Please add any image file named 'sample.jpg' to #{FILES_PATH}" unless File.file? @image_file
end

# Delete test files and delete test path 'my' on Yandex disc
def clear!
  @yd.delete('/my').should be_true if @yd
  FileUtils.rm_rf DOWNLOAD_PATH
end



spec/yandex_disk_spec.rb
#encoding: UTF-8
require 'spec_helper'
require 'fastimage'

describe YandexDisk do

  it 'should return version' do
    YandexDisk.version.should equal YandexDisk::VERSION
  end

  describe 'validate' do
    it 'authorization' do
      yd = YandexDisk.login('', '')
      expect{yd.size}.to raise_error YandexDisk::RequestError, /401 Unauthorized/
    end

    it 'wrong file' do
      yd = YandexDisk.login('', '')
      expect{yd.upload('not/exist/file')}.to raise_error YandexDisk::RequestError, /File not found/
    end

    it 'wrong path' do
      yd = YandexDisk.login('', '')
      expect{yd.create_path('/   / / / /f // / ')}.to raise_error YandexDisk::BadFormat, /Path has bad format/
    end
  end

  describe 'api' do
    before(:each) do
      @yd = YandexDisk.login(LOGIN, PWD)
    end

    describe 'size' do
      it 'in bytes' do
        size = @yd.size
        expect(size[:used] > 0 && size[:available] > 0).to be_true
      end

      it 'in readable format' do
        size = @yd.size(:h_size => true)
        mask = /Byte|KB|MB|GB/
        expect(size[:used].match(mask)[0].empty? &&
               size[:available].match(mask)[0].empty?).to be_false
      end
    end

    it 'should return list of files' do
     @yd.files.each do |item|
        [:name, :path, :created, :updated, :size, :is_file].each{|res| item[res].to_s.should_not be_empty}
     end
    end

    describe 'properties' do
      it 'for file' do
        @yd.upload(@text_file, 'my', :force => true)
        properties = @yd.properties('my/' + File.basename(@text_file))
        [:created, :updated, :type, :size, :is_file].each{|res| properties[res].to_s.should_not be_empty}
        properties[:type].should eq 'text/plain'
      end

      it 'for directory' do
        dir = '/my/photo'
        @yd.create_path(dir)
        properties = @yd.properties(dir)
        properties[:is_file].should be_false
      end
    end

    describe 'directory' do
      it 'create' do
        dir = 'my/photos/cats'
        @yd.create_path(dir).should be_true
        @yd.properties(dir)[:is_file].should be_false
      end

      it 'copy' do
        # create directories with Cyrillic name
        src = 'файлы/старые'
        des = 'файлы/здесь новые'

        @yd.create_path(src)
        @yd.create_path(des)
        # create 2 files
        files = {'file1.txt' => 'data one', 'file2.txt' => 'data two'}
        files.each do |file, text|
          new_file = File.join(FILES_PATH, file)
          File.open(new_file, 'w'){|f| f.write(text)}
          @yd.upload(new_file, src).should be_true
        end
        # copy directory
        @yd.copy(src, des)
        # check files
        @yd.files(src).size.should eq 2
        # clear
        files.each{|file, text| File.delete( File.join(FILES_PATH, file) )}
      end
    end

    describe 'file' do
      it 'upload' do
        @yd.upload(@text_file, 'my', :force => true).should be_true
        @yd.exist?('my/' + File.basename(@text_file)).should be_true
      end

      describe 'download' do
        it 'save and validate' do
          @yd.upload(@text_file, 'my', :force => true).should be_true
          path = 'my/' + File.basename(@text_file)
          @yd.download(path, DOWNLOAD_PATH).should be_true
          expect( File.read( File.join(DOWNLOAD_PATH, File.basename(@text_file)) ) ).to eq(FILE_TEXT)
        end

        it 'return and validate' do
          @yd.upload(@text_file, 'my', :force => true).should be_true
          path = 'my/' + File.basename(@text_file)
          YandexDisk::Api.any_instance.stub(:response).and_return('xxx')
          @yd.download(path).should eq FILE_TEXT
        end

        it 'unzip file' do
          file = @text_file.gsub('.txt', '.gz')
          Zlib::GzipWriter.open(file) do |gz|
            gz.write File.read(@text_file)
          end

          @yd.upload(file, 'my', :force => true).should be_true
          path = 'my/' + File.basename(file)
          @yd.should_receive(:gzip?){ true }
          @yd.download(path).should eq FILE_TEXT
        end
      end

      it 'copy' do
        @yd.upload(@text_file, 'my', :force => true)
        f_name = File.basename(@text_file)
        file = 'my/' + f_name
        new_path = 'my/text'
        @yd.create_path(new_path).should be_true
        @yd.copy(file, new_path).should be_true
        # file still exist in src path
        @yd.exist?(file).should be_true
      end

      it 'move' do
        @yd.upload(@text_file, 'my', :force => true)
        f_name = File.basename(@text_file)
        file = 'my/' + f_name
        new_path = 'my/text'
        @yd.create_path(new_path)
        @yd.move(file, new_path).should be_true
        # file not exist in src path
        @yd.exist?(file).should be_false
      end

      it 'delete' do
        @yd.upload(@text_file)
        f_name = File.basename(@text_file)
        @yd.delete(f_name).should be_true
        @yd.exist?(f_name).should be_false
      end

      it 'set public'do
        @yd.upload(@image_file, 'my', :force => true)
        # mask: http://yadi.sk/d/#############
        expect(@yd.set_public('my/' + File.basename(@image_file)) =~ /http:\/\/yadi\.sk\/d\/.+/).to be_true
      end

      it 'set private'do
        @yd.upload(@text_file, 'my', :force => true)
        f_name = 'my/' + File.basename(@text_file)
        # mask: http://yadi.sk/d/#############
        expect(@yd.set_public(f_name) =~ /http:\/\/yadi\.sk\/d\/.+/).to be_true
        expect(@yd.set_private(f_name)).to be_true
      end

      describe 'preview' do
        before(:each)do
          @yd.upload(@image_file, 'my', :force => true).should be_true
          @f_name = File.basename(@image_file)
        end

        it 'should return image in M size'do
          @yd.preview('my/' + @f_name, 'm', DOWNLOAD_PATH).should be_true
          File.file?( File.join(DOWNLOAD_PATH, @f_name) ).should be_true
        end

        it 'should return image in 300x250 size'do
          @yd.preview('my/' + @f_name, '300x250', DOWNLOAD_PATH).should be_true
          FastImage.size( File.join(DOWNLOAD_PATH, @f_name) ).should eq [300, 250]
        end

        it 'should return image in 128 pixels wide'do
          @yd.preview('my/' + @f_name, 128, DOWNLOAD_PATH).should be_true
          FastImage.size( File.join(DOWNLOAD_PATH, @f_name) )[0].should eq 128
        end
      end
    end

    it 'should catch wrong method' do
      expect { @yd.send :init_request, :none }.to raise_error YandexDisk::RequestError, /Method none not supported/
    end
  end
end

lib/yandex_disk/api.rb
require 'net/http'
require 'uri'
require 'openssl'
require 'zlib'
require 'base64'
require 'rexml/document'

require 'yandex_disk/cfg'
require 'yandex_disk/ext'
require 'yandex_disk/chunked'

module YandexDisk
  class RequestError < Exception; end
  class BadFormat < Exception; end

  class Api

    def initialize(login, pwd)
      @token = 'Basic ' + Base64.encode64("#{login}:#{pwd}")
    end

    # Example:
    #   yd.upload('/home/graph.pdf', 'my/work')
    #   => true
    #
    # Arguments:
    #   file: path to file
    #   path: path to yandex disk directory. If path not present - save to root
    #   options:
    #      chunk_size: file chunk size (default is 100)
    #      force: create path structure if not exist (raise <b>RequestError</b> if <b>path</b> not exist for default)
    def upload(file, path = '', options = {})
      # valid file?
      raise RequestError, "File not found." if file.nil? || !File.file?(file)
      # create path
      create_path(path) if options[:force]
      options[:chunk_size] ||= 1024
      @file = File.open(file)
      options[:headers] = {'Expect' => '100-continue',
                           'Transfer-Encoding' => 'chunked',
                           'content-type' => 'application/binary'}

      send_request(:put, options.merge({:path => File.join( path, File.basename(file) )}))
      @file.close
      return true
    end

    # Example:
    #   yd.download('/home/graph.pdf', '/home')
    #   => true (or file content)
    #
    # Arguments:
    #   file: path to yandex disk file
    #   save_path: path to save. If save_path not present - return file content
    def download(file, save_path = nil)
      option = {:path => file,
                :headers => {'TE' => 'chunked',
                             'Accept-Encoding' => 'gzip'}}

      send_request(:get, option)

      # unzip if zipped
      if gzip?(@response.header['Content-Encoding'])
        s_io = StringIO.new(@response.body)
        gz = Zlib::GzipReader.new(s_io)
        data = gz.read
      else
        data = @response.body
      end

      if save_path
        File.open(File.join(save_path, file.split('/').last), 'w'){|f| f.write(data)}
        return true
      else
        return data
      end
    end

    # Example:
    #   yd.create_path('/home/my/photos')
    #   => true
    #
    # Arguments:
    #   path: path to yandex disk directory hierarchy
    def create_path(path)
      c_path = ''
      path.split('/').each do |p|
        next if p.empty?
        c_path << p + '/'
        send_request(:mkcol, :path => c_path)
      end
    end
    alias_method :mkdir, :create_path

    # Example:
    #   yd.size
    #   => {:available => 312312, :used => 3123}
    # Arguments:
    #   options:
    #      readable: return size in human readable format e.g, 100K 128M 1G (false for default)
    def size(options = {})
      body = '<?xml version="1.0" encoding="utf-8"?><D:propfind xmlns:D="DAV:"><D:prop><D:quota-available-bytes/><D:quota-used-bytes/></D:prop></D:propfind>'
      send_propfind(0, :body => body)
      xml = REXML::Document.new(@response.body)
      prop = 'd:multistatus/d:response/d:propstat/d:prop/'
      available_b = xml.elements[prop + 'd:quota-available-bytes'].text.to_i
      used_b = xml.elements[prop + 'd:quota-used-bytes'].text.to_i

      return {:available => available_b.to_readable(options[:h_size]),
             :used => used_b.to_readable(options[:h_size])}
    end

    # Example:
    #   yd.exist?('/home/graph.pdf')
    #   => true
    #
    # Arguments:
    #   path: path to yandex disk directory or file
    def exist?(path)
      body = '<?xml version="1.0" encoding="utf-8"?><propfind xmlns="DAV:"><prop><displayname/></prop></propfind>'
      send_propfind(0, :path => path, :body => body)
    rescue RequestError
      return false
    end

    # Example:
    #   yd.properties('/home/graph.pdf')
    #   =>
    #       {:name => 'graph.pdf',
    #       :created => (Time),
    #       :updated => (Time),
    #       :type => 'pdf',
    #       :size => 42432,
    #       :is_file => true,
    #       :public_url => nil}
    #
    # Arguments:
    #   path: path to yandex disk directory or file
    #   h_size: return size in human readable format e.g, 100K 128M 1G (false for default)
    def properties(path, options = {})
      body = '<?xml version="1.0" encoding="utf-8"?><propfind xmlns="DAV:"><prop><displayname/><creationdate/><getlastmodified/><getcontenttype/><getcontentlength/><public_url xmlns="urn:yandex:disk:meta"/></prop></propfind>'
      send_propfind(0, :path => path, :body => body)
      prop = 'd:multistatus/d:response/d:propstat/d:prop/'
      xml = REXML::Document.new(@response.body)
      type = xml.elements[prop + 'd:getcontenttype'].text
      size = xml.elements[prop + 'd:getcontentlength'].text.to_i

      return {:name => xml.elements[prop + 'd:displayname'].text,
              :created => xml.elements[prop + 'd:getlastmodified'].text,
              :updated => xml.elements[prop + 'd:getlastmodified'].text,
              :type => type ? type : 'dir',
              :size => size.to_readable(options[:h_size]),
              :is_file => size > 0,
              :public_url => xml.elements[prop + 'public_url'].text}
    end

    # Example:
    #   yd.files('/home')
    #   =>
    #       [{:name => 'graph.pdf',
    #       :created => (Time),
    #       :updated => (Time),
    #       :type => 'pdf',
    #       :size => 42432,
    #       :is_file => true}]
    #
    # Arguments:
    #   path: path to yandex disk directory (default is <b>root</b>)
    #   root: include information of root directory or not (<b>false</b> for default)
    #   h_size: return size in human readable format e.g, 100K 128M 1G (false for default)
    def files(path = '', options = {})
      send_propfind(1, :path => path)
      xml = REXML::Document.new(@response.body)
      prop = 'd:propstat/d:prop/'
      files = []
      xml.elements.each('d:multistatus/d:response') do |res|
        name = URI.decode(res.elements[prop + 'd:displayname'].text)
        next if !options[:root] && path.split('/').last == name
        size = res.elements[prop + 'd:getcontentlength'].text.to_i rescue 0

        files << {:name => name,
                  :path => URI.decode(res.elements['d:href'].text),
                  :created => res.elements[prop + 'd:creationdate'].text,
                  :updated => res.elements[prop + 'd:getlastmodified'].text,
                  :size => size.to_readable(options[:h_size]),
                  :is_file => size > 0}
      end
      return files
    end
    alias_method :ls, :files

    # Example:
    #   yd.copy('/home/graph.pdf', 'my/work')
    #   => true
    #
    # Arguments:
    #   from: path to yandex disk directory or file
    #   to: path to yandex disk directory
    def copy(from, to)
      move_copy(:copy, from, to)
    end
    alias_method :cp, :copy

    # Example:
    #   yd.move('/home/graph.pdf', 'my/work')
    #   => true
    #
    # Arguments:
    #   from: path to yandex disk directory or file
    #   to: path to yandex disk directory
    def move(from, to)
      move_copy(:move, from, to)
    end
    alias_method :mv, :move

    # Example:
    #   yd.delete('/home/graph.pdf')
    #   => true
    #
    # Arguments:
    #   path: path to yandex disk directory or file
    def delete(path)
      send_request(:delete, :path => path)
    end
    alias_method :del, :delete

    # Example:
    #   yd.set_public('/home/graph.pdf')
    #   => http://yadi.sk/d/#############
    #
    # Arguments:
    #   path: path to yandex disk directory or file
    def set_public(path)
      body = '<propertyupdate xmlns="DAV:"><set><prop><public_url xmlns="urn:yandex:disk:meta">true</public_url></prop></set></propertyupdate>'
      send_request(:proppatch, :path => path, :body => body)
      xml = REXML::Document.new(@response.body)
      return xml.elements['d:multistatus/d:response/d:propstat/d:prop/public_url'].text
    end

    # Example:
    #   yd.set_private('/home/graph.pdf')
    #   => true
    #
    # Arguments:
    #   path: path to yandex disk directory or file
    def set_private(path)
      body = '<propertyupdate xmlns="DAV:"><remove><prop><public_url xmlns="urn:yandex:disk:meta" /></prop></remove></propertyupdate>'
      send_request(:proppatch, :path => path, :body => body)
      xml = REXML::Document.new(@response.body)
      return xml.elements['d:multistatus/d:response/d:propstat/d:prop/public_url'].text.nil?
    end

    # Example:
    #   yd.preview('/home/cat.jpg', 128, '/home/photo')
    #   => true
    #
    # Arguments:
    #   path: path to yandex disk file
    #   size: preview size (for details visit http://api.yandex.com/disk/doc/dg/reference/preview.xml)
    #   save_to: path to save
    def preview(path, size, save_to)
      send_request(:get, :path => path, :preview => size.to_s)
      File.open(File.join(save_to, path.split('/').last), 'w'){|f| f.write(@response.body)}
    end

    ########## private ##########
    private

    def gzip?(type)
      type == 'gzip'
    end

    def create_dest(from, to)
      prop = properties(from)
      to = prop[:is_file] ? to.gsub(/\/$/,'') + '/' + prop[:name] : to
      return '/' + to
    end

    def request_path(path, preview)
      if path.blank?
        return ''
      else
        path = URI.encode( path.split('/').reject{|it| it.blank?}.join('/') )
        raise Exception if path.empty?
        # image preview
        path << "?preview&size=#{preview}" if preview
        return path
      end
    rescue Exception
      raise BadFormat, 'Path has bad format.'
    end

    def init_request(method)
      case method
        when :put then Net::HTTP::Put
        when :get then Net::HTTP::Get
        when :mkcol then Net::HTTP::Mkcol
        when :propfind then Net::HTTP::Propfind
        when :copy  then Net::HTTP::Copy
        when :move  then Net::HTTP::Move
        when :delete then Net::HTTP::Delete
        when :proppatch then Net::HTTP::Proppatch
        else raise RequestError, "Method #{method} not supported."
      end
    end

    def move_copy(method, from, to)
      send_request(method, :path => from,
                           :headers => {'Destination' => create_dest(from, to)})
    end

    def send_propfind(depth, options = {})
      headers = {:headers => {'Depth' => depth.to_s,
                              'Content-Type' => 'text/xml; charset="utf-8"'}}
      send_request(:propfind, options.merge!(headers))
    end

    # return true if successful
    def send_request(method, args = {})
      # headers
      headers = {'Authorization' => @token,
                 'User-Agent' => "Ruby client library v#{YandexDisk::VERSION}"}
      headers.merge!(args[:headers]) if args[:headers]
      uri = URI.parse(YandexDisk::API_URL)
      # path
      path = uri.request_uri + request_path(args[:path], args[:preview])
      # init
      http = Net::HTTP.new(uri.host, uri.port)
      # ssl
      if uri.scheme == 'https'
        http.use_ssl = true
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
      end
      # debug
      http.set_debug_output($stderr) if YandexDisk::DEBUG
      # request
      req = init_request(method).new(path, headers)
      req.body_stream = Chunked.new(@file, args[:chunk_size]) if method == :put
      req.body = args[:body] if args[:body]
      # start
      http.start{|h| @response = h.request(req) }
      successful?(method, path)
    end

    def successful?(method, path)
      if [200, 201, 207].include?(@response.code.to_i)      ||
         @response.body.include?('resource already exists') ||
         (method == :delete && @response.body.include?('resource not found'))
        true
      else
        raise RequestError, "#{@response.code.to_i} #{@response.message}: #{@response.body} on #{path}"
      end
    end
  end

end